feat: DeepMarkRequired

Beraliv opened the pull request:
<!--
👋 Hi, thanks for sending a PR to ts-essentials! 🧡
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

- [x] Addresses an existing open issue: related to #315
- [x] Steps in [Contributing](https://github.com/ts-essentials/ts-essentials/blob/master/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

- Added `DeepMarkRequired`
    - As part of the solution, I added `LastOfUnion` and `UnionToTuple` (didn't export it to prevent unintentional usage, it is a hazardous utility type which can be broken in any TypeScript version, the mitigation - testing TypeScript versions on CI)
    - The solution intentionally accumulates mutations of required paths in place (therefore, the known limitation is the number of iterations over union elements)
    - Added `AreNonDistributiveEqual`
- Covered `DeepMarkRequired` with basic unit tests (objects-only)

som-sm commented on lib/deep-mark-required/index.ts:
```diff
@@ -0,0 +1,30 @@
+import { Prettify } from "../prettify";
+import { UnionToTuple } from "../union-to-tuple";
+
+type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
+  ? {
+      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
+        ? Prettify<RecursiveDeepMarkRequiredSinglePath<NonNullable<Type[Key]>, RestKeyPath>>
+        : Type[Key];
+    }
+  : KeyPath extends keyof Type
+  ? Prettify<MarkRequired<Type, KeyPath>>
+  : Type;
```
The current implementation fails at these cases:

#### 1. Fails to preserve `null | undefined`

```ts
type Test = RecursiveDeepMarkRequiredSinglePath<{ a: { b?: 1 } | null }, "a.b">;
//   ^? type Test = { a: { b: 1 } }
```
`Test` should have ideally been `{ a: { b: 1 } | null }` but instead it's `{ a: { b: 1 } }`. And this happens because of the usage of `NonNullable` in L11.

#### 2. Doesn't work properly with unions
```ts
type Test = RecursiveDeepMarkRequiredSinglePath<{ a: { b?: 1 } | { c: 1 } }, "a.b">;
//   ^? type Test = { a: { b?: 1 } | { c: 1 } }
```
Here, we get back the same type we passed, without "b" being marked as required. And this happens because the conditional `KeyPath extends keyof Type` fails as there are no overlapping keys between `{ b?: 1 }` and `{ c: 1 }`.

<hr />

Following is a possible fix for the above mentioned issues:
```suggestion
type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
  ? {
      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
        ? Prettify<RecursiveDeepMarkRequiredSinglePath<Type[Key], RestKeyPath>>
        : Type[Key];
    }
  : Type extends object
  ? Prettify<MarkRequired<Type, KeyPath & keyof Type>>
  : Type;
```

We can even move the conditional `Type extends object` to the top, this reads better IMO.
```ts
type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = Type extends object
  ? KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
    ? {
        [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
          ? Prettify<RecursiveDeepMarkRequiredSinglePath<Type[Key], RestKeyPath>>
          : Type[Key];
      }
    : Prettify<MarkRequired<Type, KeyPath & keyof Type>>
  : Type;
```

Beraliv replied:
@som-sm Thank you for your detailed observations!

> 1.  Fails to preserve null | undefined

I could also reproduced them after adding unit tests in https://github.com/ts-essentials/ts-essentials/pull/408/commits/3b66c318058bc111926013a3287a56f36031713c

> We can even move the conditional Type extends object to the top, this reads better IMO.

I also liked the last suggested solution, addressed it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/ca6055ce581992beb5d07a94011426aeec437ee7

Beraliv replied:
Also added unit tests with unions - https://github.com/ts-essentials/ts-essentials/pull/408/commits/e0922687ed7541ba00be72263e384d0a72164d4f

som-sm replied:
@Beraliv LGTM!

som-sm commented on lib/deep-mark-required/index.ts:
```diff
@@ -0,0 +1,30 @@
+
+type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
+  ? {
+      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
```
```suggestion
      [Key in keyof Type]: Key extends PropertyKey
```

<br />

Is there really a need for `AreNonDistributiveEqual` because neither `Key` nor `PropertyKey` would be a union here.

Beraliv replied:
Agree, addressed it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/ca6055ce581992beb5d07a94011426aeec437ee7

som-sm reviewed (commented):
@Beraliv I saw that you're planning to merge this PR, so I thought I'd give it read. I've added my feedback, let me your thoughts.

som-sm commented on lib/deep-mark-required/index.ts:
```diff
@@ -0,0 +1,30 @@
+  ? Prettify<MarkRequired<Type, KeyPath>>
+  : Type;
+
+type RecursiveDeepMarkRequired<Accumulator, KeyPaths extends string[]> = KeyPaths extends [
+  infer KeyPath,
+  ...infer RestKeyPaths
+]
+  ? RestKeyPaths extends string[]
+    ? // Keep all the changes in `Accumulator`
+      RecursiveDeepMarkRequired<RecursiveDeepMarkRequiredSinglePath<Accumulator, KeyPath>, RestKeyPaths>
+    : never
+  : Accumulator;
```
Upon further investigation, discovered that there's multiple versions of Prettier in `yarn.lock`.

<img width="889" alt="image" src="https://github.com/user-attachments/assets/67fa99ec-571a-4011-ae3d-e57e52d7ef00">

<br />
<br />

Since both `changeset` and our project allow upgrading the minor version, simply running `yarn upgrade prettier` would fix the issue.

<img width="919" alt="image" src="https://github.com/user-attachments/assets/eff76f4a-9cba-4f73-b7fb-366275ab70e4">

Beraliv replied:
This syntax appeared in TypeScript 4.7, but we still have to support 4.5 and 4.6 versions

```ts
A extends [B, infer C extends D]
              ^^^^^^^^^^^^^^^^^
```

I documented it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/98d84ef38b4d5429278dfd5c213f037b27a97f85

Beraliv replied:
@som-sm I fixed yarn.lock in https://github.com/ts-essentials/ts-essentials/commit/b34cb4913001699b4b2868164ac20966c00e6272, thanks for finding such a specific issue

som-sm replied:
@Beraliv Oh yeah, I missed that this syntax would fail in 4.5 and 4.6.

Beraliv replied:
No problems, it's a headache to remember about multiple TypeScript versions 😅 but it is a strategy of this utility library I'd like to keep

Beraliv commented on lib/deep-mark-required/index.ts:
```diff
@@ -0,0 +1,30 @@
+
+export type DeepMarkRequired<Type, KeyPathUnion extends Paths<Type>> = HasParsablePath<Type> extends false
+  ? Type
+  : RecursiveDeepMarkRequired<NonNullable<Type>, UnionToTuple<KeyPathUnion>>;
```
I agree, addressed it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/ab4f13c4038c877c045e7a54d1b72733617a99d1

Beraliv commented:
@som-sm I appreciate that you spent some time on the review. If you have time next week, would you be so kind to re-review it? I addressed all your comments, I'm happy to merge my PR

som-sm commented:
@Beraliv I've gone through the changes, and everything LGTM! Feel free to merge the PR.

Beraliv commented:
@som-sm although this change is ready, this will be scheduled for 10.1 release, so I'd like to finish all smaller changes and then I will merge this one

som-sm commented:
@Beraliv Sounds good! Let me know if there’s anything I can help with—I’d be happy to contribute.

Beraliv commented:
@som-sm FYI I'm merging this change
